File: android/src/main/java/y2k/joyreactor/tv/TagHeadersFragment.java
package y2k.joyreactor.tv;

import android.os.Bundle;
import android.support.v17.leanback.app.HeadersFragment;
import android.support.v17.leanback.app.OnHeaderViewSelectedListenerImpl;
import android.support.v17.leanback.widget.*;
import y2k.joyreactor.R;
import y2k.joyreactor.Tag;
import y2k.joyreactor.presenters.TagListPresenter;

import java.util.List;

/**
 * Created by y2k on 11/25/15.
 */
public class TagHeadersFragment extends HeadersFragment {

    TagListPresenter presenter;
    List<Tag> tags;

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        this.setOnHeaderViewSelectedListener(new OnHeaderViewSelectedListenerImpl() {

            @Override
            public void onHeaderSelected(RowHeaderPresenter.ViewHolder viewHolder, Row row) {
                if (row.getId() == 0) presenter.selectedFeatured();
                else if (row.getId() == 1) presenter.selectedFavorite();
                else presenter.selectTag(tags.get((int) (row.getId() - 2)));
            }
        });

        presenter = new TagListPresenter(new TagListPresenter.View() {

            @Override
            public void reloadData(List<Tag> tags) {
                TagHeadersFragment.this.tags = tags;
                ArrayObjectAdapter adapter = new ArrayObjectAdapter(new ListRowPresenter());

                addRow(adapter, getString(R.string.feed));
                addRow(adapter, getString(R.string.favorite));
                for (Tag tag : tags)
                    addRow(adapter, tag.title);

                setAdapter(adapter);
            }

            private void addRow(ArrayObjectAdapter adapter, String title) {
                adapter.add(new Row(adapter.size(), new HeaderItem(title)));
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.activate();
    }

    @Override
    public void onPause() {
        super.onPause();
        presenter.deactivate();
    }
}

File: android/src/main/kotlin/y2k/joyreactor/tv/TagHeadersFragment.kt
package y2k.joyreactor.tv

import android.os.Bundle
import android.support.v17.leanback.app.HeadersFragment
import android.support.v17.leanback.widget.ArrayObjectAdapter
import android.support.v17.leanback.widget.HeaderItem
import android.support.v17.leanback.widget.ListRowPresenter
import android.support.v17.leanback.widget.Row
import y2k.joyreactor.R
import y2k.joyreactor.Tag
import y2k.joyreactor.common.ServiceLocator
import y2k.joyreactor.presenters.TagListPresenter
import y2k.joyreactor.services.LifeCycleService

/**
 * Created by y2k on 11/25/15.
 */
class TagHeadersFragment : HeadersFragment() {

    lateinit var presenter: TagListPresenter
    lateinit var tags: List<Tag>
    val lifeCycleService = LifeCycleService()

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        this.setOnHeaderViewSelectedListener { viewHolder, row ->
            when (row.id) {
                0L -> presenter.selectedFeatured()
                1L -> presenter.selectedFavorite()
                else -> presenter.selectTag(tags[(row.id - 2).toInt()])
            }
        }

        presenter = ServiceLocator.provideTagListPresenter(
            object : TagListPresenter.View {

                override fun reloadData(tags: List<Tag>) {
                    this@TagHeadersFragment.tags = tags
                    val adapter = ArrayObjectAdapter(ListRowPresenter())

                    addRow(adapter, getString(R.string.feed))
                    addRow(adapter, getString(R.string.favorite))
                    for (tag in tags)
                        addRow(adapter, tag.title)

                    setAdapter(adapter)
                }

                private fun addRow(adapter: ArrayObjectAdapter, title: String?) {
                    adapter.add(Row(adapter.size().toLong(), HeaderItem(title)))
                }
            }, lifeCycleService)
    }

    override fun onResume() {
        super.onResume()
        lifeCycleService.activate()
    }

    override fun onPause() {
        super.onPause()
        lifeCycleService.deactivate()
    }
}